File: src/main/java/org/akop/ararat/graphics/CrosswordRenderer.java
// Copyright (c) 2014-2016 Akop Karapetyan
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

package org.akop.ararat.graphics;

import android.content.Context;
import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.RectF;
import android.util.DisplayMetrics;

import org.akop.ararat.core.Crossword;

import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStream;


public class CrosswordRenderer
{
	public static final int FLAG_RENDER_ANSWER  = 1;
	public static final int FLAG_RENDER_MARKERS = 2;
	public static final int FLAG_RENDER_ATTEMPT = 4;

	private static final int CELL_STROKE_COLOR = Color.parseColor("#000000");
	private static final int CELL_FILL_COLOR = Color.parseColor("#ffffff");
	private static final int PUZZLE_BG_COLOR = Color.parseColor("#000000");
	private static final int CIRCLE_STROKE_COLOR = Color.parseColor("#555555");
	private static final int TEXT_COLOR = Color.parseColor("#000000");

	private static final float CELL_STROKE_WIDTH = 1;

	private Paint mCellStrokePaint;
	private Paint mCellFillPaint;
	private Paint mPuzzleBackgroundPaint;
	private Paint mCircleStrokePaint;
	private Paint mAnswerTextPaintBase;
	private float mScaledCellStrokeWidth;

	public CrosswordRenderer(Context context)
	{
		init(context.getResources());
	}

	private void init(Resources res)
	{
		DisplayMetrics dm = res.getDisplayMetrics();

		mScaledCellStrokeWidth = CELL_STROKE_WIDTH * dm.density;

		mCellStrokePaint = new Paint();
		mCellStrokePaint.setColor(CELL_STROKE_COLOR);
		mCellStrokePaint.setStyle(Paint.Style.STROKE);
		mCellStrokePaint.setStrokeWidth(mScaledCellStrokeWidth);

		mCellFillPaint = new Paint();
		mCellFillPaint.setColor(CELL_FILL_COLOR);
		mCellFillPaint.setStyle(Paint.Style.FILL);

		mPuzzleBackgroundPaint = new Paint();
		mPuzzleBackgroundPaint.setColor(PUZZLE_BG_COLOR);
		mPuzzleBackgroundPaint.setStyle(Paint.Style.FILL);
		mPuzzleBackgroundPaint.setStrokeWidth(mScaledCellStrokeWidth);

		mCircleStrokePaint = new Paint(Paint.ANTI_ALIAS_FLAG);
		mCircleStrokePaint.setColor(CIRCLE_STROKE_COLOR);
		mCircleStrokePaint.setStyle(Paint.Style.STROKE);
		mCircleStrokePaint.setStrokeWidth(1);

		mAnswerTextPaintBase = new Paint(Paint.ANTI_ALIAS_FLAG);
		mAnswerTextPaintBase.setColor(TEXT_COLOR);
		mAnswerTextPaintBase.setTextAlign(Paint.Align.CENTER);
	}

	private void renderCell(RenderParams rp, int row, int col, RectF cellRect)
	{
		Crossword.Cell cell = rp.mCellMap[row][col];
		if (cell == null) {
			return;
		}

		rp.mCanvas.drawRect(cellRect, mCellFillPaint);
		rp.mCanvas.drawRect(cellRect, mCellStrokePaint);

		if (cell.isCircled() && (rp.mFlags & FLAG_RENDER_MARKERS) == FLAG_RENDER_MARKERS) {
			rp.mCanvas.drawCircle(cellRect.centerX(), cellRect.centerY(),
					rp.mRadius, mCircleStrokePaint);
		}

		if ((rp.mFlags & FLAG_RENDER_ANSWER) == FLAG_RENDER_ANSWER) {
			if (!cell.isEmpty()) {
				rp.mCanvas.drawText(cell.chars(), cellRect.left + rp.mCellDim / 2,
						cellRect.bottom - rp.mAnswerMetrics.descent, rp.mAnswerTextPaint);
			}
		} else if ((rp.mFlags & FLAG_RENDER_ATTEMPT) == FLAG_RENDER_ATTEMPT) {
			// FIXME: rebus
			String attempt = rp.mState.charAt(row, col);
			if (attempt != null) {
				rp.mCanvas.drawText(attempt, cellRect.left + rp.mCellDim / 2,
						cellRect.bottom - rp.mAnswerMetrics.descent, rp.mAnswerTextPaint);
			}
		}
	}

	public void renderToCanvas(Canvas canvas, Crossword crossword,
			Crossword.State state, int flags)
	{
		RenderParams rp = new RenderParams(canvas, crossword, state, flags);
		RectF puzzleRect = new RectF(0, 0, rp.mBmpW, rp.mBmpH);

		canvas.drawRect(puzzleRect, mPuzzleBackgroundPaint);

		float renderedWidth = rp.mCellDim * (float) rp.mCwW;
		float renderedHeight = rp.mCellDim * (float) rp.mCwH;

		RectF cellRect = new RectF();
		float leftmost = ((float) rp.mBmpW - renderedWidth) / 2f;
		float rectTop = ((float) rp.mBmpH - renderedHeight) / 2f;

		for (int i = 0; i < rp.mCwH; i++, rectTop += rp.mCellDim) {
			float rectLeft = leftmost;
			for (int j = 0; j < rp.mCwW; j++, rectLeft += rp.mCellDim) {
				cellRect.set(rectLeft, rectTop,
						rectLeft + rp.mCellDim, rectTop + rp.mCellDim);
				renderCell(rp, i, j, cellRect);
			}
		}
	}

	public void renderToFile(String path,
			Crossword crossword, Crossword.State state,
			int width, int height, int flags)
			throws IOException
	{
		Bitmap bmp = Bitmap.createBitmap(width, height, Bitmap.Config.RGB_565);
		renderToCanvas(new Canvas(bmp), crossword, state, flags);

		OutputStream stream = new FileOutputStream(path);
		bmp.compress(Bitmap.CompressFormat.PNG, 80, stream);
		stream.close();
	}

	private final class RenderParams
	{
		final Canvas mCanvas;
		final float mCellDim;
		final float mRadius;
		final Paint mAnswerTextPaint;
		final int mBmpW;
		final int mBmpH;
		final int mCwW;
		final int mCwH;
		final Crossword.Cell[][] mCellMap;
		final Crossword.State mState;
		final Paint.FontMetrics mAnswerMetrics;
		final int mFlags;

		RenderParams(Canvas canvas,
				Crossword crossword, Crossword.State state, int flags)
		{
			mCanvas = canvas;
			mCwW = crossword.getWidth();
			mCwH = crossword.getHeight();
			mBmpW = canvas.getWidth();
			mBmpH = canvas.getHeight();
			mFlags = flags;

			int longestCwDim = Math.max(mCwW, mCwH);
			float shortestWantedDim = Math.min(mBmpW, mBmpH);
			mCellDim = (shortestWantedDim / longestCwDim)
					- (mScaledCellStrokeWidth / longestCwDim);
			mRadius = (mCellDim / 2) - mCircleStrokePaint.getStrokeWidth();

			mAnswerTextPaint = new Paint(mAnswerTextPaintBase);
			mAnswerTextPaint.setTextSize(mCellDim * 0.75f);

			mAnswerMetrics = mAnswerTextPaint.getFontMetrics();

			mCellMap = crossword.getCellMap();
			mState = state;
		}
	}
}

File: library/src/main/java/org/akop/ararat/graphics/CrosswordRenderer.kt
// Copyright (c) Akop Karapetyan
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

package org.akop.ararat.graphics

import android.content.Context
import android.content.res.Resources
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Rect
import android.graphics.RectF

import org.akop.ararat.core.Crossword
import org.akop.ararat.core.CrosswordState

import java.io.FileOutputStream
import java.io.IOException


class CrosswordRenderer(context: Context) {

    private val cellStrokePaint = Paint()
    private val cellFillPaint = Paint()
    private val puzzleBackgroundPaint = Paint()
    private val circleStrokePaint = Paint(Paint.ANTI_ALIAS_FLAG)
    private var scaledDensity: Float = 0f
    private var cellStrokeWidthPx: Float = 0f
        set(value) {
            field = value
            cellStrokePaint.strokeWidth = value
            puzzleBackgroundPaint.strokeWidth = value
        }

    init {
        init(context.resources)
    }

    private fun init(res: Resources) {
        val dm = res.displayMetrics

        scaledDensity = dm.scaledDensity

        cellStrokePaint.color = CELL_STROKE_COLOR
        cellStrokePaint.style = Paint.Style.STROKE

        cellFillPaint.color = CELL_FILL_COLOR
        cellFillPaint.style = Paint.Style.FILL

        puzzleBackgroundPaint.color = PUZZLE_BG_COLOR
        puzzleBackgroundPaint.style = Paint.Style.FILL

        circleStrokePaint.color = CIRCLE_STROKE_COLOR
        circleStrokePaint.style = Paint.Style.STROKE

        cellStrokeWidthPx = DEFAULT_CELL_STROKE_WIDTH_DP * dm.density
    }

    private fun renderCell(rp: RenderParams, row: Int, col: Int, cellRect: RectF) {
        val cell = rp.crossword.cellMap[row][col] ?: return

        rp.canvas.drawRect(cellRect, cellFillPaint)
        rp.canvas.drawRect(cellRect, cellStrokePaint)

        if (cell.isCircled && rp.flags and FLAG_RENDER_MARKERS == FLAG_RENDER_MARKERS) {
            rp.canvas.drawCircle(cellRect.centerX(), cellRect.centerY(),
                    rp.radius, circleStrokePaint)
        }

        when {
            rp.flags and FLAG_RENDER_ANSWER == FLAG_RENDER_ANSWER -> if (!cell.isEmpty) {
                rp.canvas.drawText(cell.chars, cellRect.left + rp.cellDim / 2,
                        cellRect.bottom - rp.answerMetrics.descent, rp.answerTextPaint)
            }
            rp.flags and FLAG_RENDER_ATTEMPT == FLAG_RENDER_ATTEMPT -> {
                rp.state?.charAt(row, col)?.let {
                    val text = if (it.length > 8) it.substring(0, 8) + "…" else it
                    var textSize = rp.answerTextSize
                    var textWidth: Float

                    do {
                        rp.answerTextPaint.textSize = textSize
                        textWidth = rp.answerTextPaint.measureText(text)
                        textSize -= scaledDensity
                    } while (textWidth >= rp.answerTextSize)

                    rp.answerTextPaint.getTextBounds("A", 0, 1, rp.tempRect)
                    val xOffset = textWidth / 2f
                    val yOffset = (rp.tempRect.height() / 2).toFloat()

                    rp.canvas.drawText(text, cellRect.centerX() - xOffset,
                            cellRect.centerY() + yOffset, rp.answerTextPaint)
                }
            }
        }
    }

    fun renderToCanvas(canvas: Canvas, crossword: Crossword,
                       state: CrosswordState?, flags: Int) {
        val rp = RenderParams(canvas, crossword, state, flags)
        val puzzleRect = RectF(0f, 0f, rp.bitmapWidth.toFloat(),
                rp.bitmapHeight.toFloat())

        canvas.drawRect(puzzleRect, puzzleBackgroundPaint)

        val renderedWidth = rp.cellDim * crossword.width.toFloat()
        val renderedHeight = rp.cellDim * crossword.height.toFloat()

        val cellRect = RectF()
        val leftmost = (rp.bitmapWidth.toFloat() - renderedWidth) / 2f
        var rectTop = (rp.bitmapHeight.toFloat() - renderedHeight) / 2f

        for (i in 0 until crossword.height) {
            var rectLeft = leftmost
            for (j in 0 until crossword.width) {
                cellRect.set(rectLeft, rectTop, rectLeft + rp.cellDim, rectTop + rp.cellDim)
                renderCell(rp, i, j, cellRect)
                rectLeft += rp.cellDim
            }
            rectTop += rp.cellDim
        }
    }

    @Throws(IOException::class)
    fun renderToFile(path: String,
                     crossword: Crossword, state: CrosswordState?,
                     width: Int, height: Int, flags: Int) {
        val bmp = Bitmap.createBitmap(width, height, Bitmap.Config.RGB_565)
        renderToCanvas(Canvas(bmp), crossword, state, flags)

        FileOutputStream(path).use {
            bmp.compress(Bitmap.CompressFormat.PNG, 80, it)
        }
    }

    private inner class RenderParams(val canvas: Canvas,
                                     val crossword: Crossword,
                                     val state: CrosswordState?,
                                     val flags: Int) {

        val cellDim: Float
        val radius: Float
        val answerTextSize: Float
        val answerTextPaint = Paint(Paint.ANTI_ALIAS_FLAG)
        val bitmapWidth: Int = canvas.width
        val bitmapHeight: Int = canvas.height
        val answerMetrics: Paint.FontMetrics
        val tempRect = Rect()

        init {
            val longestCwDim = Math.max(crossword.width, crossword.height)
            val shortestWantedDim = Math.min(bitmapWidth, bitmapHeight).toFloat()
            cellDim = shortestWantedDim / longestCwDim - cellStrokeWidthPx / longestCwDim
            radius = cellDim / 2 - circleStrokePaint.strokeWidth
            answerTextSize = cellDim * 0.75f
            answerTextPaint.color = TEXT_COLOR
            answerMetrics = answerTextPaint.fontMetrics
        }
    }

    companion object {
        const val FLAG_RENDER_ANSWER  = 1
        const val FLAG_RENDER_MARKERS = 2
        const val FLAG_RENDER_ATTEMPT = 4

        private val CELL_STROKE_COLOR = Color.parseColor("#000000")
        private val CELL_FILL_COLOR = Color.parseColor("#ffffff")
        private val PUZZLE_BG_COLOR = Color.parseColor("#000000")
        private val CIRCLE_STROKE_COLOR = Color.parseColor("#555555")
        private val TEXT_COLOR = Color.parseColor("#000000")

        private const val DEFAULT_CELL_STROKE_WIDTH_DP = 1f
    }
}
